Active Directory User Lookup fails

I know we are using LDAP referrals at our site, ist by any chance not yet supported in the apache ldap client library ?

The test connection works fine, but I am unable to actually lookup any users. I double checked baseDN and search pattern, they're fine.

**Update**: Login was actually successfull, but the LDAP Settings page fails to test for users. This is not a dramatic bug, but very misleading.

```
2015-09-10T01:49:46.782+02:00 ERROR [AnyExceptionClassMapper] Unhandled exception in REST resource
java.lang.RuntimeException: ERR_02002_FAILURE_ON_UNDERLYING_CURSOR Failure on underlying Cursor.
        at org.apache.directory.api.ldap.model.cursor.CursorIterator.next(CursorIterator.java:86)
        at org.graylog2.security.ldap.LdapConnector.search(LdapConnector.java:127)
        at org.graylog2.rest.resources.system.ldap.LdapResource.testLdapConfiguration(LdapResource.java:168)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:606)
        at org.glassfish.jersey.server.model.internal.ResourceMethodInvocationHandlerFactory$1.invoke(ResourceMethodInvocationHandlerFactory.java:81)
        at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher$1.run(AbstractJavaResourceMethodDispatcher.java:144)
        at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.invoke(AbstractJavaResourceMethodDispatcher.java:161)
        at org.glassfish.jersey.server.model.internal.JavaResourceMethodDispatcherProvider$TypeOutInvoker.doDispatch(JavaResourceMethodDispatcherProvider.java:205)
        at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.dispatch(AbstractJavaResourceMethodDispatcher.java:99)
        at org.glassfish.jersey.server.model.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:389)
        at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:347)
        at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:102)
        at org.glassfish.jersey.server.ServerRuntime$2.run(ServerRuntime.java:308)
        at org.glassfish.jersey.internal.Errors$1.call(Errors.java:271)
        at org.glassfish.jersey.internal.Errors$1.call(Errors.java:267)
        at org.glassfish.jersey.internal.Errors.process(Errors.java:315)
        at org.glassfish.jersey.internal.Errors.process(Errors.java:297)
        at org.glassfish.jersey.internal.Errors.process(Errors.java:267)
        at org.glassfish.jersey.process.internal.RequestScope.runInScope(RequestScope.java:317)
        at org.glassfish.jersey.server.ServerRuntime.process(ServerRuntime.java:291)
        at org.glassfish.jersey.server.ApplicationHandler.handle(ApplicationHandler.java:1140)
        at org.graylog2.jersey.container.netty.NettyContainer.messageReceived(NettyContainer.java:356)
        at org.jboss.netty.channel.SimpleChannelUpstreamHandler.handleUpstream(SimpleChannelUpstreamHandler.java:70)
        at org.jboss.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:564)
        at org.jboss.netty.channel.DefaultChannelPipeline$DefaultChannelHandlerContext.sendUpstream(DefaultChannelPipeline.java:791)
        at org.jboss.netty.handler.execution.ChannelUpstreamEventRunnable.doRun(ChannelUpstreamEventRunnable.java:43)
        at org.jboss.netty.handler.execution.ChannelEventRunnable.run(ChannelEventRunnable.java:67)
        at com.codahale.metrics.InstrumentedExecutorService$InstrumentedRunnable.run(InstrumentedExecutorService.java:176)
        at org.jboss.netty.handler.execution.MemoryAwareThreadPoolExecutor$MemoryAwareRunnable.run(MemoryAwareThreadPoolExecutor.java:606)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
        at java.lang.Thread.run(Thread.java:745)
Caused by: org.apache.directory.api.ldap.model.cursor.CursorLdapReferralException
        at org.apache.directory.ldap.client.api.EntryCursorImpl.get(EntryCursorImpl.java:164)
        at org.apache.directory.ldap.client.api.EntryCursorImpl.get(EntryCursorImpl.java:50)
        at org.apache.directory.api.ldap.model.cursor.CursorIterator.next(CursorIterator.java:80)
        ... 34 more
```

Happens with graylog 1.2.0-4.rc.4

Ah yis. Success with rc5 for "Login Test".

Group lookup is still failing:

```
2015-09-10T12:53:34.558+02:00 ERROR [AnyExceptionClassMapper] Unhandled exception in REST resource
java.lang.RuntimeException: ERR_02002_FAILURE_ON_UNDERLYING_CURSOR Failure on underlying Cursor.
        at org.apache.directory.api.ldap.model.cursor.CursorIterator.next(CursorIterator.java:86)
        at org.graylog2.security.ldap.LdapConnector.findGroups(LdapConnector.java:219)
        at org.graylog2.security.ldap.LdapConnector.listGroups(LdapConnector.java:267)
        at org.graylog2.rest.resources.system.ldap.LdapResource.readGroups(LdapResource.java:303)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:606)
        at org.glassfish.jersey.server.model.internal.ResourceMethodInvocationHandlerFactory$1.invoke(ResourceMethodInvocationHandlerFactory.java:81)
        at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher$1.run(AbstractJavaResourceMethodDispatcher.java:144)
        at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.invoke(AbstractJavaResourceMethodDispatcher.java:161)
        at org.glassfish.jersey.server.model.internal.JavaResourceMethodDispatcherProvider$TypeOutInvoker.doDispatch(JavaResourceMethodDispatcherProvider.java:205)
        at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.dispatch(AbstractJavaResourceMethodDispatcher.java:99)
        at org.glassfish.jersey.server.model.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:389)
        at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:347)
        at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:102)
        at org.glassfish.jersey.server.ServerRuntime$2.run(ServerRuntime.java:308)
        at org.glassfish.jersey.internal.Errors$1.call(Errors.java:271)
        at org.glassfish.jersey.internal.Errors$1.call(Errors.java:267)
        at org.glassfish.jersey.internal.Errors.process(Errors.java:315)
        at org.glassfish.jersey.internal.Errors.process(Errors.java:297)
        at org.glassfish.jersey.internal.Errors.process(Errors.java:267)
        at org.glassfish.jersey.process.internal.RequestScope.runInScope(RequestScope.java:317)
        at org.glassfish.jersey.server.ServerRuntime.process(ServerRuntime.java:291)
        at org.glassfish.jersey.server.ApplicationHandler.handle(ApplicationHandler.java:1140)
        at org.graylog2.jersey.container.netty.NettyContainer.messageReceived(NettyContainer.java:356)
        at org.jboss.netty.channel.SimpleChannelUpstreamHandler.handleUpstream(SimpleChannelUpstreamHandler.java:70)
        at org.jboss.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:564)
        at org.jboss.netty.channel.DefaultChannelPipeline$DefaultChannelHandlerContext.sendUpstream(DefaultChannelPipeline.java:791)
        at org.jboss.netty.handler.execution.ChannelUpstreamEventRunnable.doRun(ChannelUpstreamEventRunnable.java:43)
        at org.jboss.netty.handler.execution.ChannelEventRunnable.run(ChannelEventRunnable.java:67)
        at com.codahale.metrics.InstrumentedExecutorService$InstrumentedRunnable.run(InstrumentedExecutorService.java:176)
        at org.jboss.netty.handler.execution.MemoryAwareThreadPoolExecutor$MemoryAwareRunnable.run(MemoryAwareThreadPoolExecutor.java:606)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
        at java.lang.Thread.run(Thread.java:745)
Caused by: org.apache.directory.api.ldap.model.cursor.CursorLdapReferralException
        at org.apache.directory.ldap.client.api.EntryCursorImpl.get(EntryCursorImpl.java:164)
        at org.apache.directory.ldap.client.api.EntryCursorImpl.get(EntryCursorImpl.java:50)
        at org.apache.directory.api.ldap.model.cursor.CursorIterator.next(CursorIterator.java:80)
        ... 35 more

```

--- a/graylog2-plugin-interfaces/src/main/java/org/graylog2/plugin/DocsHelper.java
+++ b/graylog2-plugin-interfaces/src/main/java/org/graylog2/plugin/DocsHelper.java
@@ -25,7 +25,8 @@ package org.graylog2.plugin;
 
 public enum DocsHelper {
     PAGE_SENDING_JSONPATH("sending_data.html#json-path-from-http-api-input"),
-    PAGE_ES_CONFIGURATION("configuring_es.html");
+    PAGE_ES_CONFIGURATION("configuring_es.html"),
+    PAGE_LDAP_TROUBLESHOOTING("users_roles.html#troubleshooting");
 
     private static final String DOCS_URL = "http://docs.graylog.org/en/";
 

--- a/graylog2-server/src/main/java/org/graylog2/rest/resources/system/ldap/LdapResource.java
+++ b/graylog2-server/src/main/java/org/graylog2/rest/resources/system/ldap/LdapResource.java
@@ -306,7 +306,7 @@ public class LdapResource extends RestResource {
             );
         } catch (IOException | LdapException e) {
             LOG.error("Unable to retrieve available LDAP groups", e);
-            throw new InternalServerErrorException(e);
+            throw new InternalServerErrorException("Unable to retrieve available LDAP groups", e);
         }
     }
 }

--- a/graylog2-server/src/main/java/org/graylog2/rest/resources/system/ldap/LdapResource.java
+++ b/graylog2-server/src/main/java/org/graylog2/rest/resources/system/ldap/LdapResource.java
@@ -298,15 +298,13 @@ public class LdapResource extends RestResource {
             config.setCredentials(ldapSettings.getSystemPassword());
         }
 
-        try {
-            LdapNetworkConnection connection = ldapConnector.connect(config);
-            final Set<String> groups = ldapConnector.listGroups(connection,
-                                                                ldapSettings.getGroupSearchBase(),
-                                                                ldapSettings.getGroupSearchPattern(),
-                                                                ldapSettings.getGroupIdAttribute()
+        try (LdapNetworkConnection connection = ldapConnector.connect(config)) {
+            return ldapConnector.listGroups(connection,
+                                            ldapSettings.getGroupSearchBase(),
+                                            ldapSettings.getGroupSearchPattern(),
+                                            ldapSettings.getGroupIdAttribute()
             );
-            return groups;
-        } catch (LdapException e) {
+        } catch (IOException | LdapException e) {
             LOG.error("Unable to retrieve available LDAP groups", e);
             throw new InternalServerErrorException(e);
         }

--- a/graylog2-server/src/main/java/org/graylog2/security/ldap/LdapConnector.java
+++ b/graylog2-server/src/main/java/org/graylog2/security/ldap/LdapConnector.java
@@ -33,7 +33,9 @@ import org.apache.directory.api.ldap.model.message.ResultCodeEnum;
 import org.apache.directory.api.ldap.model.message.SearchScope;
 import org.apache.directory.ldap.client.api.LdapConnectionConfig;
 import org.apache.directory.ldap.client.api.LdapNetworkConnection;
+import org.graylog2.plugin.DocsHelper;
 import org.graylog2.shared.security.ldap.LdapEntry;
+import org.graylog2.shared.utilities.ExceptionUtils;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
@@ -205,7 +207,7 @@ public class LdapConnector {
                                   String groupSearchBase,
                                   String groupSearchPattern,
                                   String groupIdAttribute,
-                                  String dn) throws LdapException {
+                                  String dn) {
         final Set<String> groups = Sets.newHashSet();
 
         EntryCursor groupSearch = null;
@@ -251,6 +253,12 @@ public class LdapConnector {
                     }
                 }
             }
+        } catch (Exception e) {
+            LOG.warn(
+                    "Unable to iterate over user's groups, unable to perform group mapping. Graylog does not support " +
+                            "LDAP referrals at the moment. Please see " +
+                            DocsHelper.PAGE_LDAP_TROUBLESHOOTING.toString() + " for more information.",
+                    ExceptionUtils.getRootCause(e));
         } finally {
             if (groupSearch != null) {
                 groupSearch.close();
@@ -263,7 +271,7 @@ public class LdapConnector {
     public Set<String> listGroups(LdapNetworkConnection connection,
                                   String groupSearchBase,
                                   String groupSearchPattern,
-                                  String groupIdAttribute) throws LdapException {
+                                  String groupIdAttribute) {
         return findGroups(connection, groupSearchBase, groupSearchPattern, groupIdAttribute, null);
     }
 

